File: genui/02_ag_ui/step_01_ag_ui_server/sse.mjs
// Server-sent events, the reading side. fetch() gives a byte stream; this
// turns it into JSON events. One event is the `data:` lines up to a blank
// line. AG-UI puts one whole JSON event on each data line.

export async function* readEvents(response) {
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true }).replace(/\r\n/g, "\n"); // the spec allows either ending
    let cut;
    while ((cut = buffer.indexOf("\n\n")) >= 0) {
      const block = buffer.slice(0, cut);
      buffer = buffer.slice(cut + 2);
      const event = parseBlock(block);
      if (event) yield event;
    }
  }
  const last = parseBlock(buffer);
  if (last) yield last;
}

export function parseBlock(block) {
  const data = block
    .split("\n")
    .filter((line) => line.startsWith("data:"))
    .map((line) => line.slice(5).trimStart());
  return data.length ? JSON.parse(data.join("\n")) : null;
}

File: genui/02_ag_ui/step_02_ag_ui_tools_and_state/sse.py
"""Server-sent events, the reading side, in Python. The page does the same in
sse.mjs; the tests and demo.py use this copy.

One event is one or more `data:` lines followed by a blank line. AG-UI puts
the whole JSON event on a single data line.
"""

import json


def parse_sse(text):
    """Split raw SSE text into the JSON objects it carries, in order."""
    events = []
    for block in text.replace("\r\n", "\n").split("\n\n"):  # the spec allows either line ending
        data = [line[5:].lstrip() for line in block.splitlines() if line.startswith("data:")]
        if data:
            events.append(json.loads("\n".join(data)))
    return events
